fem/nonlininteg.cpp

```diff
@@ -385,4 +439,230 @@ void HyperelasticNLFIntegrator::AssembleElementGrad(const FiniteElement &el,
+                 " has finite element space of incorrect block number");
+   }
+
+   int dof_u = el[0]->GetDof();
+   int dof_p = el[1]->GetDof();
+
+   int dim = el[0]->GetDim();
+
+   DSh_u.SetSize(dof_u, dim);
+   DS_u.SetSize(dof_u, dim);
+   J0i.SetSize(dim);
+   F.SetSize(dim);
+   FinvT.SetSize(dim);
+   P.SetSize(dim);
+   PMatI_u.UseExternalData(elfun[0]->GetData(), dof_u, dim);
+   elvec[0]->SetSize(dof_u*dim);
+   PMatO_u.UseExternalData(elvec[0]->GetData(), dof_u, dim);
+
+   Sh_p.SetSize(dof_p);
+   elvec[1]->SetSize(dof_p);
+
+   int intorder = 2*el[0]->GetOrder() + 3; // <---
+   const IntegrationRule &ir = IntRules.Get(el[0]->GetGeomType(), intorder);
+
+   *elvec[0] = 0.0;
+   *elvec[1] = 0.0;
+
+   for (int i = 0; i < ir.GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      Tr.SetIntPoint(&ip);
+      CalcInverse(Tr.Jacobian(), J0i);
+
+      el[0]->CalcDShape(ip, DSh_u);
+      Mult(DSh_u, J0i, DS_u);
+      MultAtB(PMatI_u, DS_u, F);
+
+      el[1]->CalcShape(ip, Sh_p);
+
+      double pres = Sh_p * *elfun[1];
+      double mu = c_mu->Eval(Tr, ip);
+      double dJ = F.Det();
+
+      CalcInverseTranspose(F, FinvT);
+
+      P = 0.0;
+      P.Add(mu * dJ, F);
+      P.Add(-1.0 * pres * dJ, FinvT);
+      P *= ip.weight*Tr.Weight();
+
+      AddMultABt(DS_u, P, PMatO_u);
+
+      elvec[1]->Add(ip.weight * Tr.Weight() * (dJ - 1.0), Sh_p);
+   }
+
+}
+
+void IncompressibleNeoHookeanIntegrator::AssembleElementGrad(
+   const Array<const FiniteElement*> &el,
+   ElementTransformation &Tr,
+   const Array<const Vector *> &elfun,
+   const Array2D<DenseMatrix *> &elmats)
+{
+   int dof_u = el[0]->GetDof();
+   int dof_p = el[1]->GetDof();
+
+   int dim = el[0]->GetDim();
+
+   elmats(0,0)->SetSize(dof_u*dim, dof_u*dim);
+   elmats(0,1)->SetSize(dof_u*dim, dof_p);
+   elmats(1,0)->SetSize(dof_p, dof_u*dim);
+   elmats(1,1)->SetSize(dof_p, dof_p);
+
+   *elmats(0,0) = 0.0;
+   *elmats(0,1) = 0.0;
+   *elmats(1,0) = 0.0;
+   *elmats(1,1) = 0.0;
+
+   DSh_u.SetSize(dof_u, dim);
+   DS_u.SetSize(dof_u, dim);
+   J0i.SetSize(dim);
+   F.SetSize(dim);
+   FinvT.SetSize(dim);
+   Finv.SetSize(dim);
+   P.SetSize(dim);
+   PMatI_u.UseExternalData(elfun[0]->GetData(), dof_u, dim);
+   Sh_p.SetSize(dof_p);
+
+   int intorder = 2*el[0]->GetOrder() + 3; // <---
+   const IntegrationRule &ir = IntRules.Get(el[0]->GetGeomType(), intorder);
+
+   for (int i = 0; i < ir.GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      Tr.SetIntPoint(&ip);
+      CalcInverse(Tr.Jacobian(), J0i);
+
+      el[0]->CalcDShape(ip, DSh_u);
+      Mult(DSh_u, J0i, DS_u);
+      MultAtB(PMatI_u, DS_u, F);
+
+      el[1]->CalcShape(ip, Sh_p);
+      double pres = Sh_p * *elfun[1];
+      double mu = c_mu->Eval(Tr, ip);
+      double dJ = F.Det();
+      double dJ_FinvT_DS;
+
+      CalcInverseTranspose(F, FinvT);
+
+      // u,u block
+      for (int i_u = 0; i_u < dof_u; i_u++)
```

Comment:
It doesn't matter for integers these days so I wouldn't change it here, but its a good habit to use the ++i operator rather than the i++ operator for esoteric performance reasons.

Reply:
Good to know! I always remember one is better, but I can never remember which is which.